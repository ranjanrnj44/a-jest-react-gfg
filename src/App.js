import { createContext, useState } from 'react';
import './App.css';
import Button from './components/Button';
import TestElements from './components/TestElements';
import CounterComponent from './components/featureCounter/CounterComponent';
import NoCounterContext from './components/NoCounterContext';
import CounterContext from './components/CounterContext';

// context
export const CounterContextCreation = createContext();

function App() {
  // STATE :: initial state
  const [toggle, setToggle] = useState(false);
  // action :: handleToggle
  const handleToggle = () => setToggle(prev => !prev);
  // counter state
  const [counter, setCounter] = useState(0);
  // context counter state
  const [contextCounter, setContextCounter] = useState(0);
  
  // non context contextIncrementHandler, contextDecrementHandler
  const incrementHandler = () => setCounter(prev => prev + 1);
  const decrementHandler = () => setCounter(prev => prev - 1);
  
  // contextIncrementHandler, contextDecrementHandler
  const contextIncrementHandler = () => setContextCounter(prev => prev + 1);
  const contextDecrementHandler = () => setContextCounter(prev => prev - 1);

  // context creation
  return (
    <div>
      <h1>Hello World!</h1>
      <TestElements />
      <br />
      <Button toggle={toggle} setToggle={handleToggle} />
      <br />
      <CounterComponent />
      <br />
      <NoCounterContext counter={counter} incrementHandler={incrementHandler} decrementHandler={decrementHandler} />
      <br />
      <CounterContextCreation.Provider value={{contextCounter, contextIncrementHandler, contextDecrementHandler}}>
      <CounterContext />
      </CounterContextCreation.Provider>
    </div>
  );
}

export default App;
