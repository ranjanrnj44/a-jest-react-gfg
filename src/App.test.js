import { screen, render, cleanup, fireEvent } from "@testing-library/react";
import App from "./App";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import counterReducer from "./components/featureCounter/CounterSlice";

// cleans up after each test case
afterEach(cleanup);

// create a store
  const store = configureStore({
    reducer: {
      counter: counterReducer,
    },
  });

// for App component
describe("App component loads by default", () => {
  test("should show heading of the App component", () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(screen.getByText(/Hello World!/i)).toBeInTheDocument();
  });
  test.skip("should match the snapshot and the component", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(asFragment(<App />)).toMatchSnapshot();
  });
});

// button component
describe("Button component", () => {
  test("should show and hide the text if button is clicked", () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(
      screen.queryByText(/Ranjan is writing UT using JEST!!!/i)
    ).not.toBeInTheDocument();
    const button = screen
      .getByTestId("show-hide-button");
    fireEvent.click(button);
    expect(
      screen.getByText(/Ranjan is writing UT using JEST!!!/i)
    ).toBeInTheDocument();
    fireEvent.click(button);
    expect(
      screen.queryByText(/Ranjan is writing UT using JEST!!!/i)
    ).not.toBeInTheDocument();
  });
});
