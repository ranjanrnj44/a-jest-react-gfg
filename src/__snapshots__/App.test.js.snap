// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component loads by default should match the snapshot and the component 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Hello World!
    </h1>
    <h1
      data-testid="counter"
    >
      0
    </h1>
    <button
      data-testid="button-up"
    >
      Increment
    </button>
    <button
      data-testid="button-disabled"
      disabled=""
    >
      Disabled
    </button>
    <button
      data-testid="button-down"
    >
      Decrement
    </button>
    <br />
    <button
      name="Button"
    >
      Button
    </button>
    <br />
    <h1>
      RTK counter component!
    </h1>
    <h3>
       count is : 0
    </h3>
    <button
      data-testid="redux-increment"
    >
      Increment
    </button>
    <button
      data-testid="redux-decrement"
    >
      Decrement
    </button>
    <button
      data-testid="redux-incrementBy"
    >
      IncrementBy
    </button>
  </div>
</DocumentFragment>
`;

exports[`should render App component by default should match the snapshot and the component 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Hello World!
    </h1>
    <button>
      Button
    </button>
  </div>
</DocumentFragment>
`;
